File: backend/app/services/backtest_engine.py
"""
Simple backtest engine.
Strategies are rule-based; replace with InStock engine calls in production.
"""

import asyncio
import hashlib
import math
import random
from datetime import timedelta

from app.models import (
    BacktestRequest,
    BacktestResult,
    BacktestStats,
    BacktestStatus,
    BacktestStrategy,
    EquityPoint,
    Trade,
)

STRATEGIES: list[BacktestStrategy] = [
    BacktestStrategy(
        id="macd_cross",
        name="MACD 金叉",
        description="MACD DIF 上穿 DEA 买入，下穿卖出",
        markets=["CN", "US"],
    ),
    BacktestStrategy(
        id="rsi_reversal",
        name="RSI 超卖反弹",
        description="RSI<30 买入，RSI>70 卖出",
        markets=["CN", "US"],
    ),
    BacktestStrategy(
        id="boll_breakout",
        name="布林带突破",
        description="收盘价突破上轨买入，跌破下轨卖出",
        markets=["CN", "US"],
    ),
    BacktestStrategy(
        id="kdj_signal",
        name="KDJ 信号",
        description="KDJ J线超卖回升买入",
        markets=["CN"],
    ),
    BacktestStrategy(
        id="volume_surge",
        name="放量突破",
        description="成交量放大 2 倍且价格创新高买入",
        markets=["CN", "US"],
    ),
]


def _build_seed(req: BacktestRequest) -> int:
    payload = "|".join(
        [
            req.strategy_id,
            ",".join(sorted(req.tickers)),
            req.start_date.isoformat(),
            req.end_date.isoformat(),
            f"{req.initial_capital:.2f}",
        ]
    )
    return int(hashlib.sha256(payload.encode("utf-8")).hexdigest()[:16], 16)


def _simulate_backtest(req: BacktestRequest) -> BacktestResult:
    """Deterministic simulation based on strategy + tickers + dates."""
    rng = random.Random(_build_seed(req))
    days = (req.end_date - req.start_date).days
    if days < 10:
        raise ValueError("回测区间至少需要 10 天")

    capital = req.initial_capital
    benchmark_capital = req.initial_capital
    equity_curve: list[EquityPoint] = []
    trades: list[Trade] = []

    for i in range(0, days, max(1, days // 100)):
        point_date = req.start_date + timedelta(days=i)
        daily_ret = rng.uniform(-0.015, 0.02)
        bench_ret = rng.uniform(-0.01, 0.015)
        capital *= 1 + daily_ret
        benchmark_capital *= 1 + bench_ret
        equity_curve.append(
            EquityPoint(
                date=point_date,
                equity=round(capital, 2),
                benchmark=round(benchmark_capital, 2),
            )
        )

    for ticker in req.tickers:
        n_trades = rng.randint(3, 8)
        for _ in range(n_trades):
            trade_date = req.start_date + timedelta(days=rng.randint(1, days - 1))
            price = rng.uniform(10, 500)
            shares = round(rng.uniform(100, 1000), 0)
            pnl = round(price * shares * rng.uniform(-0.1, 0.2), 2)
            trades.append(
                Trade(
                    ticker=ticker,
                    action=rng.choice(["BUY", "SELL"]),
                    date=trade_date,
                    price=round(price, 2),
                    shares=shares,
                    pnl=pnl,
                )
            )

    trades.sort(key=lambda trade: trade.date)

    final_equity = equity_curve[-1].equity if equity_curve else capital
    total_return = (final_equity - req.initial_capital) / req.initial_capital * 100
    years = max(days / 365, 0.01)
    ann_return = ((final_equity / req.initial_capital) ** (1 / years) - 1) * 100

    returns = []
    for i in range(1, len(equity_curve)):
        prev = equity_curve[i - 1].equity
        curr = equity_curve[i].equity
        returns.append((curr - prev) / prev)
    avg_r = sum(returns) / len(returns) if returns else 0
    std_r = math.sqrt(sum((ret - avg_r) ** 2 for ret in returns) / len(returns)) if returns else 1
    sharpe = avg_r / std_r * math.sqrt(252) if std_r > 0 else 0

    peak = req.initial_capital
    max_dd = 0.0
    for point in equity_curve:
        if point.equity > peak:
            peak = point.equity
        drawdown = (peak - point.equity) / peak * 100
        if drawdown > max_dd:
            max_dd = drawdown

    win_trades = [trade for trade in trades if trade.pnl > 0]
    win_rate = len(win_trades) / len(trades) * 100 if trades else 0

    stats = BacktestStats(
        total_return_pct=round(total_return, 2),
        annualized_return_pct=round(ann_return, 2),
        sharpe_ratio=round(sharpe, 3),
        max_drawdown_pct=round(max_dd, 2),
        win_rate_pct=round(win_rate, 1),
        total_trades=len(trades),
    )

    return BacktestResult(
        job_id="",
        status=BacktestStatus.DONE,
        strategy_id=req.strategy_id,
        tickers=req.tickers,
        start_date=req.start_date,
        end_date=req.end_date,
        equity_curve=equity_curve,
        trades=trades,
        stats=stats,
    )


async def run_backtest_async(req: BacktestRequest, job_id: str) -> BacktestResult:
    await asyncio.sleep(0.1)
    try:
        result = _simulate_backtest(req)
        result.job_id = job_id
        return result
    except Exception as exc:
        return BacktestResult(
            job_id=job_id,
            status=BacktestStatus.FAILED,
            strategy_id=req.strategy_id,
            tickers=req.tickers,
            start_date=req.start_date,
            end_date=req.end_date,
            error=str(exc),
        )


def get_strategies() -> list[BacktestStrategy]:
    return STRATEGIES

File: server/services/backtest-engine.ts
import { createHash } from 'crypto'
import type {
  BacktestRequest, BacktestResult, BacktestStats,
  BacktestStrategy, EquityPoint, Trade,
} from '../types/index.js'

export const STRATEGIES: BacktestStrategy[] = [
  { id: 'macd_cross',    name: 'MACD 金叉',    description: 'MACD DIF 上穿 DEA 买入，下穿卖出',  markets: ['CN', 'US'] },
  { id: 'rsi_reversal',  name: 'RSI 超卖反弹', description: 'RSI<30 买入，RSI>70 卖出',          markets: ['CN', 'US'] },
  { id: 'boll_breakout', name: '布林带突破',    description: '收盘价突破上轨买入，跌破下轨卖出', markets: ['CN', 'US'] },
  { id: 'kdj_signal',    name: 'KDJ 信号',      description: 'KDJ J线超卖回升买入',              markets: ['CN'] },
  { id: 'volume_surge',  name: '放量突破',      description: '成交量放大 2 倍且价格创新高买入',  markets: ['CN', 'US'] },
]

const STRATEGY_IDS = new Set(STRATEGIES.map((s) => s.id))

export function isValidStrategy(id: string): boolean {
  return STRATEGY_IDS.has(id)
}

export function getStrategies(): BacktestStrategy[] {
  return STRATEGIES
}

function buildSeed(req: BacktestRequest): number {
  const payload = [
    req.strategy_id,
    [...req.tickers].sort().join(','),
    req.start_date,
    req.end_date,
    ((req.initial_capital ?? 100_000)).toFixed(2),
  ].join('|')
  return parseInt(createHash('sha256').update(payload, 'utf8').digest('hex').slice(0, 16), 16)
}

function makePrng(seed: number) {
  let s = seed >>> 0
  return {
    random(): number {
      s = (Math.imul(s, 1664525) + 1013904223) >>> 0
      return s / 0x100000000
    },
    uniform(lo: number, hi: number): number {
      return lo + this.random() * (hi - lo)
    },
    randint(lo: number, hi: number): number {
      return Math.floor(this.uniform(lo, hi + 1))
    },
    choice<T>(arr: T[]): T {
      return arr[Math.floor(this.random() * arr.length)]
    },
  }
}

function addDays(dateStr: string, days: number): string {
  const d = new Date(dateStr)
  d.setDate(d.getDate() + days)
  return d.toISOString().slice(0, 10)
}

function dateDiffDays(a: string, b: string): number {
  return Math.round((new Date(b).getTime() - new Date(a).getTime()) / 86_400_000)
}

export function simulateBacktest(req: BacktestRequest): BacktestResult {
  const rng = makePrng(buildSeed(req))
  const days = dateDiffDays(req.start_date, req.end_date)
  if (days < 10) throw new Error('回测区间至少需要 10 天')

  const initialCapital = req.initial_capital ?? 100_000
  let capital = initialCapital
  let benchmarkCapital = initialCapital
  const equityCurve: EquityPoint[] = []
  const trades: Trade[] = []

  const step = Math.max(1, Math.floor(days / 100))
  for (let i = 0; i < days; i += step) {
    const pointDate = addDays(req.start_date, i)
    const dailyRet = rng.uniform(-0.015, 0.02)
    const benchRet = rng.uniform(-0.01, 0.015)
    capital *= 1 + dailyRet
    benchmarkCapital *= 1 + benchRet
    equityCurve.push({
      date: pointDate,
      equity: Math.round(capital * 100) / 100,
      benchmark: Math.round(benchmarkCapital * 100) / 100,
    })
  }

  for (const ticker of req.tickers) {
    const nTrades = rng.randint(3, 8)
    for (let t = 0; t < nTrades; t++) {
      const tradeDate = addDays(req.start_date, rng.randint(1, days - 1))
      const price = rng.uniform(10, 500)
      const shares = Math.round(rng.uniform(100, 1000))
      const pnl = Math.round(price * shares * rng.uniform(-0.1, 0.2) * 100) / 100
      trades.push({
        ticker,
        action: rng.choice(['BUY', 'SELL'] as const),
        date: tradeDate,
        price: Math.round(price * 100) / 100,
        shares,
        pnl,
      })
    }
  }

  trades.sort((a, b) => a.date.localeCompare(b.date))

  const finalEquity = equityCurve.length > 0 ? equityCurve[equityCurve.length - 1].equity : capital
  const totalReturn = (finalEquity - initialCapital) / initialCapital * 100
  const years = Math.max(days / 365, 0.01)
  const annReturn = ((finalEquity / initialCapital) ** (1 / years) - 1) * 100

  const returns: number[] = []
  for (let i = 1; i < equityCurve.length; i++) {
    const prev = equityCurve[i - 1].equity
    const curr = equityCurve[i].equity
    returns.push((curr - prev) / prev)
  }
  const avgR = returns.length > 0 ? returns.reduce((a, b) => a + b, 0) / returns.length : 0
  const variance = returns.length > 0 ? returns.reduce((s, r) => s + (r - avgR) ** 2, 0) / returns.length : 1
  const stdR = Math.sqrt(variance) || 1
  const sharpe = avgR / stdR * Math.sqrt(252)

  let peak = initialCapital
  let maxDD = 0
  for (const pt of equityCurve) {
    if (pt.equity > peak) peak = pt.equity
    const dd = (peak - pt.equity) / peak * 100
    if (dd > maxDD) maxDD = dd
  }

  const winTrades = trades.filter((t) => t.pnl > 0)
  const winRate = trades.length > 0 ? winTrades.length / trades.length * 100 : 0

  const stats: BacktestStats = {
    total_return_pct: Math.round(totalReturn * 100) / 100,
    annualized_return_pct: Math.round(annReturn * 100) / 100,
    sharpe_ratio: Math.round(sharpe * 1000) / 1000,
    max_drawdown_pct: Math.round(maxDD * 100) / 100,
    win_rate_pct: Math.round(winRate * 10) / 10,
    total_trades: trades.length,
  }

  return {
    job_id: '',
    status: 'DONE',
    strategy_id: req.strategy_id,
    tickers: req.tickers,
    start_date: req.start_date,
    end_date: req.end_date,
    equity_curve: equityCurve,
    trades,
    stats,
    created_at: new Date().toISOString(),
  }
}

const _jobs = new Map<string, BacktestResult>()

export function getJob(jobId: string): BacktestResult | undefined {
  return _jobs.get(jobId)
}

export async function runBacktestAsync(req: BacktestRequest, jobId: string): Promise<void> {
  _jobs.set(jobId, {
    job_id: jobId, status: 'RUNNING',
    strategy_id: req.strategy_id, tickers: req.tickers,
    start_date: req.start_date, end_date: req.end_date,
    equity_curve: [], trades: [], stats: null,
    created_at: new Date().toISOString(),
  })
  Promise.resolve().then(() => {
    try {
      const result = simulateBacktest(req)
      result.job_id = jobId
      _jobs.set(jobId, result)
    } catch (err) {
      _jobs.set(jobId, {
        job_id: jobId, status: 'FAILED',
        strategy_id: req.strategy_id, tickers: req.tickers,
        start_date: req.start_date, end_date: req.end_date,
        equity_curve: [], trades: [], stats: null,
        error: err instanceof Error ? err.message : String(err),
        created_at: new Date().toISOString(),
      })
    }
  })
}
